horreum-client/src/main/java/io/hyperfoil/tools/auth/KeycloakClientRequestFilter.java

```diff
@@ -50,24 +57,23 @@ public KeycloakClientRequestFilter(String keycloakBaseUrl,
 	}
 
 	@Override
-	public void filter(ClientRequestContext requestContext) {
+	public void filter(ClientRequestContext requestContext) throws IOException {
 		try {
-			final String accessToken = getAccessToken();
-			requestContext.getHeaders().add(HttpHeaders.AUTHORIZATION, BEARER_SCHEME_WITH_SPACE + accessToken);
+			requestContext.getHeaders().add(HttpHeaders.AUTHORIZATION, BEARER_SCHEME_WITH_SPACE.concat(getAccessToken()));
+			if (showAuthMethod) {
+				LOG.infov("Authentication with OIDC token");
+				showAuthMethod = false;
+			}
 		} catch (Exception ex) {
-			LOG.warnf(ex, "Access token is not available", ex);
-			// TODO: 401 should be only on 401 response
-			if (ex.getCause() instanceof HttpHostConnectException) {
-				LOG.warnf("Aborting the request with HTTP 500 error");
-				requestContext.abortWith(Response.status(500).build());
-			} else {
-				LOG.warnf("Aborting the request with HTTP 401 error");
-				requestContext.abortWith(Response.status(401).build());
+			if (showAuthMethod) {
+				LOG.infov("Using Basic authentication as OIDC server replied with {0}", ex.getMessage());
+				showAuthMethod = false;
 			}
+			basicAuthentication.filter(requestContext);
```

Comment:
I am confused about this fallback here, in `HorreumClient` there is a check for keycloak being configured `if (keycloakConfig.url == null || keycloakConfig.url.isEmpty())`, and then either `BasicAuthentication` or `KeycloakClientRequestFilter` is registered with the client builder.   

Here, if there is an exception, it is reverting to basicAuth.  

If a user configures keycloak, and the authentication fails, shouldn't we just error here?

Reply:
these are slightly different. `keycloakConfig.url` may not exist if authentication is handled by Horreum (without a keycloak instance) and in that case it makes no sense to try OIDC authentication.  

if the OIDC is out of our control, then the client will try to authenticate with and fallback to basic after it gets the `400` error from the server. this way there are no changes in the way clients are configured. 

at the moment the clients use the same `api/config/keycloak` endpoint the UI uses, and because of that can't know if the OIDC server is under keycloak's control or not. we could also improve the output of that endpoint to inform the clients which auth method to use.

Reply:
scratch that last comment. the client may have to use more that one authentication scheme to authenticate to a single Horreum instance. it may have to authenticate as a user or as a machine account. one example of that is the integration test !! 

an alternative is to add a parameter to configure the client (to select between `machine-account` and `user-account`), having one of those methods as default. this can be a breaking change though.

Reply:
I don't get the question. The test currently uses the client with the `horreum.bootstrap` user account (OIDC) to setup a machine account (Basic) . Machine accounts always use Basic scheme, while user accounts can be either OIDC or Basic. 

We can use two or more test profiles, but the bottom line is that the client needs to support both schemes at the same time.

Reply:
I think this is ok for now, we might want to revisit when we remove implement API keys and possibly remove machine accounts